### PR TITLE
Exclusions, reports, and various bug fixes

### PR DESCRIPTION
**Pull request by ken-duck**
A rather large PR which provides the implementation for the new (more advanced) exclusions support, as well as JUnit-friendly reports. There are some bug fixes in here as well.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,10 +102,67 @@ If the `--info` flag is provided to gradle it will output a dependency tree whic
...
 
-TODO
+The gradle plugin supports writing out test reports in the correct XML format for the
+[Jenkins JUnit Reporting Plugin](https://wiki.jenkins.io/display/JENKINS/JUnit+Plugin).
+To switch on this reporting, set the path to the report in you project's build.gradle file using the **"junitReport"** element like so:
+
+```
+audit {
+        failOnError = false
+        ignore = [ 'ch.qos.logback:logback-core' ]        ]
```
Comment (MarkusAmshove) [inline README.md]: There seems to be an extra `]`

File: README.md [hunk 2]
```diff
@@ -102,10 +102,67 @@ If the `--info` flag is provided to gradle it will output a dependency tree whic
...
+    sh "[[ ! \$(grep '<failure' ./ossindex/junitReport.xml) ]]"
+```
+
+Ensures that the build fails if any failures are reported.
+
+Set
+```
+    failOnError = false
+```
+As failOnError is true by default and will cause the scan to exit on the first failure, instead of finding them all.
```
Comment (MarkusAmshove) [inline README.md]: Doesn't the `post` and `always` step act like a `finally` block, so this shouldn't be needed?

I use `try { runTests } finally { junit… }` in my builds to always record test results but still fail the build if a test failed, because the Exception bubbles up. 
I'm not familar with the `post always` semantic though

File: src/main/java/net/ossindex/gradle/OssIndexPlugin.java [hunk 3]
```diff
@@ -45,27 +52,74 @@ public void apply(Project project) {
...
     }
 
+    /**
+     * Get the proxy if it exists. Check 3 different places:
+     *
+     *   1. The build.gradle file
+     *   2. The local properties
+     *   3. The system properties
+     */
     private Proxy getProxy(Project project, String scheme) {
```
Comment (MarkusAmshove) [inline src/main/java/net/ossindex/gradle/OssIndexPlugin.java]: It seems like the system properties gets the highest precedence in defining a proxy, because the `proxy` is `new`ed in every if branch, is this intended?
I would expect the `build.gradle` to have the highest precedence, then the local and then system.

That means that if I specify all 3, I want to get the one from the `build.gradle`.

But maybe it is intended this way :-)

File: src/main/java/net/ossindex/gradle/audit/DependencyAuditor.java [hunk 4]
```diff
@@ -59,20 +95,34 @@ private void addArtifactsToAudit(Set<GradleArtifact> gradleArtifacts) {
...
-    parents.put(parent, null);
-    gradleArtifact.getAllChildren().forEach(c -> addPackageDependencies(parent, c));
+  private PackageCoordinate buildCoordinate(final GradleArtifact gradleArtifact) {
+    PackageCoordinate coord = PackageCoordinate.newBuilder()
+        .withFormat("maven")
+        .withNamespace(gradleArtifact.getGroup())
+        .withName(gradleArtifact.getName())
+        .withVersion(gradleArtifact.getVersion())
+        .build();
+    return coord;
```
Comment (MarkusAmshove) [inline src/main/java/net/ossindex/gradle/audit/DependencyAuditor.java]: `return PackageCoordinate.newBuilder().....build();`

File: src/main/java/net/ossindex/gradle/OssIndexPlugin.java (hunk 3 above)
Comment (MarkusAmshove) [reply src/main/java/net/ossindex/gradle/OssIndexPlugin.java]: Oh I didn't catch the `!proxy.isValid()` in the other cases. Makes sense now :)

**Comment by ken-duck** [comment]
Thanks for the feedback. Sorry for the delay it has been a busy couple of weeks. I expect to provide an update in the next couple of days.

**Comment by ken-duck** [comment]
BTW, thanks very much for the extra eyes, Markus. Very appreciated. I would like to point out this PR is a collaboration between myself and @museadmin , who has been invaluable not just in the coding but also in providing a significant amount of testing support.

Also a side note, I am working on a new PR which will hopefully be done this week, which will add OSS Index 3.0 API support. This new API provides access to a newer database of vulnerabilities, which has much more data already and has more vulnerabilities added on a near-daily basis.

File: README.md (hunk 2 above)
Comment (museadmin) [reply README.md]: Hi Markus,

It is necessary to explicitly set this false as it's true by default. This has the effect of failing the build immediately, which was a problem for me. I tested this on a rather large and ungainly Java project. When the gradle audit was run, it would split the audit out into multiple, 20+, sub projects. Each of which would then have its own sub audit. 

So any sub audits running after a failure would be skipped and we would not see the results.

The work to include this in the pipeline also saw me adding support for the Jenkins JUnit plugin which gives a nice report on all of the failures. 

So in summary, when in a pipeline, this needs to be false!

Brad

Comment (MarkusAmshove) [reply README.md]: Ah, I see :)

**Comment by ken-duck** [comment]
Finally found a gap in the schedule and made the changes. I have also almost completed the 3.0 integration. I'll have a PR up for that shortly. I'm not going to give a time cause I will probably end up lying again :|

Soon...
